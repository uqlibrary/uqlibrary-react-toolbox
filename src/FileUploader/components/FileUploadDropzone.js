import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import FileUploadDropzoneStaticContent from './FileUploadDropzoneStaticContent';

export default class FileUploadDropzone extends PureComponent {
    static propTypes = {
        onDrop: PropTypes.func.isRequired,
        maxSize: PropTypes.number.isRequired,
        locale: PropTypes.object.isRequired,
        fileNameRestrictions: PropTypes.instanceOf(RegExp).isRequired,
        filesInQueue: PropTypes.array,
        fileUploadLimit: PropTypes.number,
        disabled: PropTypes.bool
    };

    static defaultProps = {
        fileUploadLimit: 10,
        filesInQueue: []
    };

    constructor(props) {
        super(props);
        this.dropzoneRef = null;
    }

    /**
     * Try to read file and set error for a folder
     *
     * @param file
     * @param errors
     * @param resolve
     */
    readFile = (file, errors, resolve) => {
        const fileReader = new FileReader();
        fileReader.onerror = () => {
            errors.push(file.name);
            return resolve(false);
        };
        fileReader.onload = () => resolve(file);
        const slice = file.slice(0, 10);
        return fileReader.readAsDataURL(slice);
    };

    /**
     * Remove duplicate files from given accepted files
     *
     * @param incomingFiles
     * @param filesInQueue - list of names of files in queue
     * @returns Object
     */
    removeDuplicate = (incomingFiles, filesInQueue) => {
        // Ignore files from incomingFiles which are already in files queue
        const uniqueFiles = incomingFiles.filter(file => filesInQueue.indexOf(file.name) === -1);
        const duplicateFiles = incomingFiles.filter(file => filesInQueue.indexOf(file.name) >= 0).map(file => file.name);

        // Return unique files and errors with duplicate file names
        return {uniqueFiles: uniqueFiles, duplicateFiles: duplicateFiles};
    };

    /**
     * Remove folders from the list
     *
     * @param filesAndFolders files and/or folders
     * @param errors
     * @returns {Promise.<*>}
     */
    removeDroppedFolders = (filesAndFolders, errors) => {
        return Promise.all(
            filesAndFolders.map(file => {
                return new Promise(resolve => {
                    this.readFile(file, errors, resolve);
                });
            })
        );
    };

    /**
     * Remove invalid file names
     *
     * @param incomingFiles - array of files
     * @param fileNameRestrictions - RegExp
     * @returns Object
     */
    removeInvalidFileNames = (incomingFiles, fileNameRestrictions) => {
        const validFiles = incomingFiles
            .filter(file => (file && new RegExp(fileNameRestrictions, 'gi').test(file.name)));
        const invalidFileNames = incomingFiles
            .filter(file => (file && !(new RegExp(fileNameRestrictions, 'gi').test(file.name))))
            .map(file => file.name);

        return {validFiles: validFiles, invalidFileNames: invalidFileNames};
    };

    /**
     * Remove files if there are too many files
     *
     * @param incomingFiles - array of files
     * @param maxAllowed files to return
     * @returns Object
     */
    removeTooManyFiles = (incomingFiles, maxAllowed) => {
        const tooManyFiles = incomingFiles.slice(maxAllowed).map(file => file.name);
        const limitedFiles = incomingFiles.slice(0, maxAllowed);

        return {limitedFiles: limitedFiles, tooManyFiles: tooManyFiles};
    };

    /**
     * Handle accepted and rejected files on dropped in Dropzone
     *
     * @param incomingFiles
     * @param rejectedFiles
     * @private
     */
    _onDrop = (incomingFiles, rejectedFiles) => {
        const {fileNameRestrictions, filesInQueue, fileUploadLimit} = this.props;
        const notFiles = [];

        // Remove folders from accepted files (async)
        this.removeDroppedFolders([...incomingFiles], notFiles)
            .then(onlyFiles => {
                // Remove invalid file names
                const {validFiles, invalidFileNames} = this.removeInvalidFileNames(onlyFiles, fileNameRestrictions);

                // Remove duplicate files from accepted files
                const {uniqueFiles, duplicateFiles} = this.removeDuplicate(validFiles, filesInQueue);

                // Remove files exceeding the max number of files allowed
                const {limitedFiles, tooManyFiles} = this.removeTooManyFiles(uniqueFiles, fileUploadLimit - filesInQueue.length);

                this.props.onDrop(
                    limitedFiles.map(file => ({fileData: file, name: file.name, size: file.size})),
                    {
                        tooBigFiles: rejectedFiles.map(file => file.name),
                        notFiles: notFiles,
                        invalidFileNames: invalidFileNames,
                        duplicateFiles: duplicateFiles,
                        tooManyFiles: tooManyFiles
                    }
                );
            });
    };

    /**
     * Open dropzone on key pressed
     */
    _onKeyPress = () => {
        this.dropzoneRef.open();
    };

    render() {
        const {maxSize, disabled, locale} = this.props;
        return (
            <div>
                <div className="columns">
                    <div className="column"  tabIndex="0" onKeyPress={this._onKeyPress}>
                        <Dropzone
                            ref={(ref) => {this.dropzoneRef = ref;}}
                            maxSize={maxSize}
                            onDrop={this._onDrop}
                            style={{padding: '0px'}}
                            disabled={disabled}
                            disableClick={disabled}
                            disablePreview>
                            <FileUploadDropzoneStaticContent locale={locale}/>
                        </Dropzone>
                    </div>
                </div>
            </div>
        );
    }
}
