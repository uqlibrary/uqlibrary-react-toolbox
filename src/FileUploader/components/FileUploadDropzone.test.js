import FileUploadDropzone from './FileUploadDropzone';

function setup(testProps, isShallow = true) {
    const props = {
        onDrop: jest.fn(),
        locale: {},
        maxSize: 1000,
        fileNameRestrictions: /.+/,
        ...testProps,
    };
    return getElement(FileUploadDropzone, props, isShallow);
}

describe('Component FileUploadDropzone', () => {
    let getMockFile;

    beforeEach(() => {
        const _File = window.File;
        const FILE = (data = [''], name) => new _File(data, name, {lastModified: 12345678912});
        window.File = jest.fn((data, name) => FILE(data, name));
        getMockFile = (name) => new File([''], name);
    });

    it('should render component with default props', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render disabled component', () => {
        const wrapper = setup({disabled: true}, false);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should open files selection dialog', () => {
        const wrapper = setup({}, false);
        expect(toJson(wrapper)).toMatchSnapshot();

        const testFn = jest.fn();

        wrapper.find('FileUploadDropzone').instance().dropzoneRef.open = testFn;
        wrapper.find('FileUploadDropzone').instance()._onKeyPress();

        wrapper.update();
        expect(testFn).toHaveBeenCalled();
    });

    it('should remove duplicate files', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('d.txt')];
        const {uniqueFiles, duplicateFiles} = wrapper.instance().removeDuplicate(files, ['a.txt', 'b.txt', 'c.txt']);

        expect(uniqueFiles.length).toEqual(1);
        expect(duplicateFiles.length).toEqual(1);
    });

    it('should not remove any files if there are no duplicate files', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('d.txt')];
        const {uniqueFiles, duplicateFiles} = wrapper.instance().removeDuplicate(files, ['a.txt', 'b.txt']);

        expect(uniqueFiles.length).toEqual(2);
        expect(duplicateFiles.length).toEqual(0);
    });

    it('should not remove any files if there are no files', () => {
        const wrapper = setup({});

        const files = [];
        const {uniqueFiles, duplicateFiles} = wrapper.instance().removeDuplicate(files, ['a.txt', 'b.txt']);

        expect(uniqueFiles.length).toEqual(0);
        expect(duplicateFiles.length).toEqual(0);
    });

    it('should remove files with invalid names', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('1.txt')];
        const {validFiles, invalidFileNames} = wrapper.instance().removeInvalidFileNames(files, /^[a-z].+/);

        expect(validFiles.length).toEqual(1);
        expect(invalidFileNames.length).toEqual(1);
    });

    it('should not remove any files if there are no invalid names of files', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('a.txt')];
        const {validFiles, invalidFileNames} = wrapper.instance().removeInvalidFileNames(files, /^[a-z].+/);

        expect(validFiles.length).toEqual(2);
        expect(invalidFileNames.length).toEqual(0);
    });

    it('should not remove any files if there are no files supplied', () => {
        const wrapper = setup({});

        const files = [];
        const {validFiles, invalidFileNames} = wrapper.instance().removeInvalidFileNames(files, /^[a-z].+/);

        expect(validFiles.length).toEqual(0);
        expect(invalidFileNames.length).toEqual(0);
    });

    it('should remove files exceeding max allowed number of files in removeTooManyFiles', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('1.txt'), getMockFile('1a.txt')];
        const {limitedFiles, tooManyFiles} = wrapper.instance().removeTooManyFiles(files, 2);

        expect(limitedFiles.length).toEqual(2);
        expect(tooManyFiles.length).toEqual(1);
    });

    it('should not remove any files if number doesn\'t exceed max allowed number of files in removeTooManyFiles', () => {
        const wrapper = setup({});

        const files = [getMockFile('c.txt'), getMockFile('1.txt'), getMockFile('1a.txt')];
        const {limitedFiles, tooManyFiles} = wrapper.instance().removeTooManyFiles(files, 4);

        expect(limitedFiles.length).toEqual(3);
        expect(tooManyFiles.length).toEqual(0);
    });

    it('should not remove any files if there are no files supplied to removeTooManyFiles', () => {
        const wrapper = setup({});

        const files = [];
        const {limitedFiles, tooManyFiles} = wrapper.instance().removeTooManyFiles(files, 3);

        expect(limitedFiles.length).toEqual(0);
        expect(tooManyFiles.length).toEqual(0);
    });

    it('should filter folders out from the file list in removeDroppedFolders', async () => {
        const wrapper = setup({});

        const file_a = getMockFile('a.txt');
        file_a.slice = (x, y) => true;
        const file_b = getMockFile('droppedFolder');
        file_b.slice = (x, y) => false;
        const file_c = getMockFile('c.txt');
        file_c.slice = (x, y) => true;

        const accepted = [file_a, file_b, file_c];
        const errors = [];

        wrapper.instance().readFile = jest.fn((file, errors, resolve) => {
            file.slice() ? resolve(file) : resolve(false);
        });

        await expect(wrapper.instance().removeDroppedFolders(accepted, errors)).resolves.toEqual([file_a, false, file_c]);
    });

    it('should set all error messages', async () => {
        const file_a = getMockFile('a.txt');
        const file_b = getMockFile('b.txt');
        const file_b_dup = getMockFile('b.txt');
        const file_c = getMockFile('c.txt');
        const file_d = getMockFile('web_d.txt');
        const file_e = getMockFile('e.txt');
        const file_f = getMockFile('f.txt');
        const file_g = getMockFile('g.txt');
        const onDropTestFn = jest.fn();

        const wrapper = setup({
            fileUploadLimit: 4,
            filesInQueue: [file_a.name, file_b.name],
            onDrop: onDropTestFn,
            fileNameRestrictions: /^(?=^\S*$)(?=^[^\.]+\.[^\.]+$)(?=.{1,45}$)(?!(web_|preview_|thumbnail_|stream_|fezacml_|presmd_))[a-z][a-z\d\-_\.]+/
        });

        const expectedFiles = [file_c, file_f].map(file => ({fileData: file, name: file.name, size: file.size}));
        const expectedError = {
            tooBigFiles: ['e.txt'],
            notFiles: [],
            invalidFileNames: ['web_d.txt'],
            duplicateFiles: ['b.txt'],
            tooManyFiles: ['g.txt']
        };

        const accepted = [file_b_dup, file_c, file_d, file_f, file_g];
        wrapper.instance().removeDroppedFolders = jest.fn((accepted, {}) => new Promise(resolve => resolve([file_b_dup, file_c, file_d, file_f, file_g])));

        await wrapper.instance()._onDrop(accepted, [file_e]);
        // wrapper.update();
        expect(onDropTestFn).toHaveBeenCalledWith(expectedFiles, expectedError);
    });
});
