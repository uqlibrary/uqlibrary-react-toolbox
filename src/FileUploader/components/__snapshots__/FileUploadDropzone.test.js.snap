// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component FileUploadDropzone should open files selection dialog 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    }
  }
>
  <FileUploadDropzone
    fileNameRestrictions={/\\.\\+/}
    fileUploadLimit={10}
    filesInQueue={Array []}
    locale={Object {}}
    maxSize={1000}
    onDrop={[MockFunction]}
  >
    <div>
      <div
        className="columns"
      >
        <div
          className="column"
          onKeyPress={[Function]}
          tabIndex="0"
        >
          <t
            disableClick={false}
            disablePreview={true}
            disabled={false}
            maxSize={1000}
            minSize={0}
            multiple={true}
            onDrop={[Function]}
            preventDropOnDocument={true}
            style={
              Object {
                "padding": "0px",
              }
            }
          >
            <div
              aria-disabled={false}
              className=""
              onClick={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              style={
                Object {
                  "padding": "0px",
                }
              }
            >
              <FileUploadDropzoneStaticContent
                locale={Object {}}
              >
                <div
                  className="columns file-instructions"
                >
                  <div
                    className="column"
                  />
                  <div
                    className="column upload-instructions"
                  >
                    <FontIcon
                      className="material-icons"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="material-icons"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        style={
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "display": "inline-block",
                            "fontSize": 24,
                            "muiPrepared": true,
                            "position": "relative",
                            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                            "userSelect": "none",
                          }
                        }
                      >
                        cloud_upload
                      </span>
                    </FontIcon>
                  </div>
                </div>
              </FileUploadDropzoneStaticContent>
              <input
                autoComplete="off"
                disabled={false}
                multiple={true}
                onChange={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="file"
              />
            </div>
          </t>
        </div>
      </div>
    </div>
  </FileUploadDropzone>
</Provider>
`;

exports[`Component FileUploadDropzone should render component with default props 1`] = `
<div>
  <div
    className="columns"
  >
    <div
      className="column"
      onKeyPress={[Function]}
      tabIndex="0"
    >
      <t
        disableClick={false}
        disablePreview={true}
        disabled={false}
        maxSize={1000}
        minSize={0}
        multiple={true}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "padding": "0px",
          }
        }
      >
        <FileUploadDropzoneStaticContent
          locale={Object {}}
        />
      </t>
    </div>
  </div>
</div>
`;

exports[`Component FileUploadDropzone should render disabled component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    }
  }
>
  <FileUploadDropzone
    disabled={true}
    fileNameRestrictions={/\\.\\+/}
    fileUploadLimit={10}
    filesInQueue={Array []}
    locale={Object {}}
    maxSize={1000}
    onDrop={[MockFunction]}
  >
    <div>
      <div
        className="columns"
      >
        <div
          className="column"
          onKeyPress={[Function]}
          tabIndex="0"
        >
          <t
            disableClick={true}
            disablePreview={true}
            disabled={true}
            maxSize={1000}
            minSize={0}
            multiple={true}
            onDrop={[Function]}
            preventDropOnDocument={true}
            style={
              Object {
                "padding": "0px",
              }
            }
          >
            <div
              aria-disabled={true}
              className=""
              onClick={null}
              onDragEnter={null}
              onDragLeave={null}
              onDragOver={null}
              onDragStart={null}
              onDrop={null}
              style={
                Object {
                  "padding": "0px",
                }
              }
            >
              <FileUploadDropzoneStaticContent
                locale={Object {}}
              >
                <div
                  className="columns file-instructions"
                >
                  <div
                    className="column"
                  />
                  <div
                    className="column upload-instructions"
                  >
                    <FontIcon
                      className="material-icons"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="material-icons"
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        style={
                          Object {
                            "color": "rgba(0, 0, 0, 0.87)",
                            "display": "inline-block",
                            "fontSize": 24,
                            "muiPrepared": true,
                            "position": "relative",
                            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                            "userSelect": "none",
                          }
                        }
                      >
                        cloud_upload
                      </span>
                    </FontIcon>
                  </div>
                </div>
              </FileUploadDropzoneStaticContent>
              <input
                autoComplete="off"
                disabled={true}
                multiple={true}
                onChange={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="file"
              />
            </div>
          </t>
        </div>
      </div>
    </div>
  </FileUploadDropzone>
</Provider>
`;
